src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs

```diff
@@ -78,53 +80,73 @@ public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs
-                                    model.GetDeclarationDiagnostics(span: null, cancellationToken: cancellationToken));
+                await AnalyzeForKind(document, AnalysisKind.Semantic, cancellationToken).ConfigureAwait(false);
 
-                Debug.Assert(document.Project.Solution.Workspace == _workspace);
+                bool IsSemanticAnalysisOn()
+                {
+                    // right now, there is no way to observe diagnostics for closed file.
+                    if (!_workspace.IsDocumentOpen(document.Id))
+                    {
+                        return false;
+                    }
+
+                    if (_workspace.Options.GetOption(InternalRuntimeDiagnosticOptions.Semantic))
+                    {
+                        return true;
+                    }
+
+                    return _workspace.Options.GetOption(InternalRuntimeDiagnosticOptions.ScriptSemantic) && document.SourceCodeKind == SourceCodeKind.Script;
+                }
+            }
 
-                var diagnosticData = diagnostics == null ? ImmutableArray<DiagnosticData>.Empty : diagnostics.Select(d => DiagnosticData.Create(document, d)).ToImmutableArrayOrEmpty();
+            private async Task AnalyzeForKind(Document document, AnalysisKind kind, CancellationToken cancellationToken)
+            {
+                var diagnosticData = await _service._analyzerService.GetDiagnosticsAsync(document, GetAnalyzers(), kind, cancellationToken).ConfigureAwait(false);
 
                 _service.RaiseDiagnosticsUpdated(
-                    DiagnosticsUpdatedArgs.DiagnosticsCreated(new DefaultUpdateArgsId(_workspace.Kind, Semantic, document.Id),
-                    _workspace, document.Project.Solution, document.Project.Id, document.Id, diagnosticData));
+                    DiagnosticsUpdatedArgs.DiagnosticsCreated(new DefaultUpdateArgsId(_workspace.Kind, kind, document.Id),
+                    _workspace, document.Project.Solution, document.Project.Id, document.Id, diagnosticData.ToImmutableArrayOrEmpty()));
+
+                IEnumerable<DiagnosticAnalyzer> GetAnalyzers()
+                {
+                    // C# or VB document that supports compiler
+                    var compilerAnalyzer = _service._analyzerService.GetCompilerDiagnosticAnalyzer(document.Project.Language);
+                    if (compilerAnalyzer != null)
+                    {
+                        return SpecializedCollections.SingletonEnumerable(compilerAnalyzer);
+                    }
+
+                    // document that doesn't support compiler diagnostics such as fsharp or typescript
+                    return _service._analyzerService.GetDiagnosticAnalyzers(document.Project);
+                }
             }
 
             public void RemoveDocument(DocumentId documentId)
             {
                 // a file is removed from misc project
```

Comment:
> // a file is removed from misc project [](start = 15, length = 39)

Is this comment still relevant? This shouldn't have any knowledge of misc workspace, correct?

Reply:
Is ScriptSemantic missing here?

---
In reply to: [238470136](https://github.com/dotnet/roslyn/pull/31134#discussion_r238470136) [](ancestors = 238470136)